### DISCUSSION TITLE
Why was BitmapPool deprecated

### DISCUSSION THREAD

**Discussion by sudeshim3** [coil-kt/coil#1186] [General]
So, I was using BitmapPool until 1.x.x version. However in 2.x.x version BitmapPool is removed. Can someone explain why it was removed and what are the alternatives

**Comment by colinrtwhite** [comment]
`BitmapPool` and `PoolableViewTarget` have been removed from the library in 2.x. Bitmap pooling was removed because:

- It's most effective on <= API 23 and has become less effective with newer Android releases.

- Removing bitmap pooling allows Coil to use immutable bitmaps, which have performance benefits.
- There's runtime overhead to manage the bitmap pool.
- Bitmap pooling creates design restrictions on Coil's API as it requires tracking if a bitmap is eligible for pooling. Removing bitmap pooling allows Coil to expose the result `Drawable` in more places (e.g. `Listener`, `Disposable`). Additionally, this means Coil doesn't have to clear `ImageView`s, which has can cause [issues](https://github.com/coil-kt/coil/issues/650).
- Bitmap pooling is [error-prone](https://github.com/coil-kt/coil/issues/546). Allocating a new bitmap is much safer than attempting to re-use a bitmap that could still be in use.

Also Android's [`ImageDecoder`](https://developer.android.com/reference/android/graphics/ImageDecoder) (which is supposed to replace [`BitmapFactory`](https://developer.android.com/reference/android/graphics/BitmapFactory)) intentionally doesn't support bitmap pooling due to the above reasons.

If you need `BitmapPool` you can copy [the implementation](https://github.com/coil-kt/coil/tree/1.x/coil-base/src/main/java/coil/bitmap) into your project.
